scribus/scpainter.cpp

```diff
@@ -1983,9 +1999,27 @@ void ScPainter::drawText(QRectF area, QString text, bool filled, int align)
 	cairo_move_to (m_cr, x, y);
 	m_stroke.getRgbF(&r, &g, &b);
 	cairo_set_source_rgba( m_cr, r, g, b, m_stroke_trans );
+
 	for (int a = 0; a < textList.count(); ++a)
 	{
-		cairo_show_text (m_cr, textList[a].toUtf8());
+		StoryText story;
+		story.insertChars(textList[a]);
+		CharStyle charStyle(m_font, m_fontSize);
+		story.setCharStyle(0, textList[a].count(), charStyle);
+		TextShaper textShaper(item, story, 0);
+		QList<GlyphRun> glyphRun = textShaper.shape();
+
+		QString text1 = textList[a];
+		cairo_glyph_t cairo_glyphs[text1.count()];
+		for (int i=0; i < glyphRun.size(); i++)
+		{
+			cairo_glyphs[i].index = glyphRun[i].glyphs().first().glyph;
+			cairo_glyphs[i].x = x + (glyphRun[i].glyphs().first().xoffset/100.);
+			cairo_glyphs[i].y = y - (glyphRun[i].glyphs().first().yoffset/100.);
+			x += glyphRun[i].glyphs().first().xadvance/100.;
+			y -= glyphRun[i].glyphs().first().yadvance/100.;
+		}
```

Comment:
This code does not make much sense. `TextShaper::shape()` returns a list of `GlyphRuns`; you should loop over that list then loop over the glyphs of each run and append those to Cairo glyphs. Also the length of Cairo glyphs array is not the same as text length. It should be something like:

````c++
QList<GlyphRun> glyphRuns = textShaper.shape();
QVector<cairo_glyph_t> cairoGlyphs;
foreach (const GlyphRun &run, glyphRuns)
{
        foreach (const GlyphLayout &gl, run.glyphs())
        {
               cairo_glyph_t g;
               g.x = x + gl.xoffset;
               g.y = y - g.yoffset;
               x += gl.xadvance
               cairoGlyphs.append(g);
        }
}

cairo_show_glyphs(m_cr, cairoGlyphs.data(), cairoGlyphs..count());
````